latitudesh/latitudesh-go
# chore: refactor servers and virtual_networks tests

#### What does this PR do?
This PR aims to verify all of our existing endpoints to prepare for our terraform provider `v1.0.0` launch.

Most of the refactoring done by this PR is organizing the tests in subtests. This way it's easier to verify where a test is failing. Also, there were some adjustments to some delete test functions.

#### Description of Task to be completed
Verify and refactor tests in order to guarantee all the SDK functionality.

*Verified endpoints:*
- [x] Servers
- [x] Virtual Networks
- [x] Virtual Networks Assignments
- [x] Operating Systems
- [x] Plans
- [x] Regions
- [x] SSH Keys
- [x] Teams
- [x] Team Members
- [x] User Data 

#### How to test?
To run the tests, use the following command:
```
LATITUDE_AUTH_TOKEN=<API-TOKEN> LATITUDE_TEST_ACTUAL_API=true LATITUDE_TEST_RECORDER=<RECORD-MODE> go test -run "^<Test-NAME>$"
```
`<API-TOKEN>` : Your API token
`<RECORD-MODE>` : can be `record` (will record the interactions with the api and save in a fixtures folder) or `play` (will play the recorded interactions instead of doing the actual requests to the api)
`<Test-NAME>` : the Name of the test function that you want to run (you can verify them at the `*_tests.go` files) 

⚠️ Attention
- When running all the tests at the same time, some of them might fail because of a bug (golang tries to create the same project more than once) I'll investigate it and open another PR solving this issue. At the time being the only way to have a clean test is to run each one on it's own.
- As the tests run against production, `TestAccServerBasic` can fail due to stock levels.
- Unfortunately, running `TestAccTeamBasic` will create another team in production (and we don't have an endpoint to delete teams from what i've seen). I'm studying the possibility of mocking some of the API calls so we don't span the API.
